### DISCUSSION TITLE
Assertion failing without retrying

### DISCUSSION THREAD

**Discussion by motykadaw** [cypress-io/cypress#24642] [Questions and Help]
I have a custom function that finds the table (list) content and returns it as an object:
```

Cypress.Commands.add(
  "getTableContents",
  { prevSubject: ["element"] },
  (subject, rowSelector, valueGetters) => {
    if (subject.get().length > 1) {
      throw new Error(
        `Selector "${subject.selector}" returned more than 1 element.`
      );
    }

    return cy
      .testEl(rowSelector)
      .then(($nodes) =>
        $nodes
          .get()
          .map((row) =>
            mapObjIndexed((valueGetter) => valueGetter(row), valueGetters)
          )
      );
  }
);
```
([mapObjIndexed](https://ramdajs.com/docs/#mapObjIndexed)).

I'm using in the following way (simplified for readability):
```
(...) // test code that adds the page with proper name
  cy
    .testEl("pages-list")
    .getTableContents("table-row", {
      name: testValueGetter("name"),
    })
    .should("containSubset", [{ name: "Test Page Name" }]);
```  

With
```
export const testValueGetter = (selector: string) => (element: HTMLElement) => {
  const value = Cypress.$(element).find(`[data-test="${selector}"]`).get(0);
  return trim(value?.textContent || "");
};
```

Initially the function correctly resolves the list of values, but adding new page is asynchronous so it appears some time after the assertion has already failed as it doesn't wait, it just fails the test as soon as it resolves the initial list state. Any changes in timeout configs don't help, not sure how to solve that. Any tips please?

**Comment by portyu9** [comment]
Hi @motykadaw 

The `.then()` inside `getTableContents()` is what breaks retryability.

`.then()` runs once. Your command therefore converts the DOM into an array once, returns that array, and the later `.should()` keeps seeing that same value even if the page updates.

Increasing the timeout does not help because Cypress has no retriable query left to execute.

Move the transformation into a `.should(callback)` attached directly to the DOM query:

```js
cy.get('[data-test="table-row"]').should(($rows) => {
  const contents = [...$rows].map((row) => ({
    name: testValueGetter("name")(row),
  }))

  expect(contents).to.containSubset([
    { name: "Test Page Name" },
  ])
})
```

Cypress will now re-query the DOM and re-run the callback until the assertion passes or times out.

For a reusable abstraction on current Cypress, another option is `Cypress.Commands.addQuery()`. Custom queries were added specifically for synchronous, idempotent operations that need Cypress retryability.

The key point is: **don't turn the current DOM into a non-query value with `.then()` before the assertion that needs to wait for an update.**

Please mark this answer as accepted if it helped, thank you.
